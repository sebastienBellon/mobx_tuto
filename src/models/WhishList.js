import { types } from "mobx-state-tree";
import Reactotron from "reactotron-react-js";
import { values } from "mobx";

export const WhishListItem = types
  .model({
    name: types.string,
    price: types.number,
    image: types.optional(types.string, ""),
  })
  .actions((self) => ({
    changeName(newName) {
      self.name = newName;
    },
    changePrice(newPrice) {
      self.price = newPrice;
    },
  }));

export const WishList = types
  .model({
    items: types.optional(types.array(WhishListItem), []),
  })
  .actions((self) => ({
    add(item) {
      self.items.push(item);
    },
  }));

// export const Todo = types
//   .model({
//     name: types.optional(types.string, ""),
//     done: types.optional(types.boolean, false),
//   })
//   .actions((self) => ({
//     setName(newName) {
//       self.name = newName;
//     },

//     toggle() {
//       self.done = !self.done;
//     },
//   }));

// export const User = types.model({
//   name: types.optional(types.string, ""),
// });

// const RootStore = types
//   .model({
//     users: types.map(User),
//     todos: types.optional(types.map(Todo), {}),
//   })
//   .views((self) => ({
//     get pendingCount() {
//       return values(self.todos).filter((todo) => !todo.done).length;
//     },
//     get completedCount() {
//       return values(self.todos).filter((todo) => todo.done).length;
//     },
//     getTodosWhereDoneIs(done) {
//       return values(self.todos).filter((todo) => todo.done === done);
//     },
//   }))
//   .actions((self) => ({
//     addTodo(id, name) {
//       self.todos.set(id, Todo.create({ name }));
//     },
//   }));

// export const store = RootStore.create({
//   users: {}, // users is not required really since arrays and maps are optional by default since MST3
// });

// // let reactotron-mst know about it
// Reactotron.trackMstNode(store);
