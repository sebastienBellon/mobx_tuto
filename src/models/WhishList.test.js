import { WhishListItem, WishList } from "./WhishList";
import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";

it("can create an instance of a model", () => {
  const item = WhishListItem.create({
    name: "Chronicles of seb",
    price: 28.73,
    image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
  });

  expect(item.price).toBe(28.73);
  item.changeName("test");
  expect(item.name).toBe("test");
});

it("can create a whishList", () => {
  const list = WishList.create({
    items: [
      {
        name: "Chronicles of seb",
        price: 28.73,
        image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
      },
      {
        name: "Chronicles of seb",
        price: 28.73,
        image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
      },
    ],
  });
  expect(list.items.length).toBe(2);
  expect(list.items[0].price).toBe(28.73);
});

it("can add new items", () => {
  const list = WishList.create();
  const states = [];
  onSnapshot(list, (snapshot) => {
    states.push(snapshot);
  });

  list.add({
    name: "Chronicles of seb",
    price: 28.73,
    image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
  });
  list.add({
    name: "Chronicles of seb",
    price: 28.73,
    image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
  });

  expect(list.items.length).toBe(2);
  list.add(
    WhishListItem.create({
      name: "test",
      price: 28.73,
      image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
    })
  );
  expect(list.items.length).toBe(3);
  expect(list.items[2].name).toBe("test");

  expect(getSnapshot(list)).toEqual({
    items: [
      {
        name: "Chronicles of seb",
        price: 28.73,
        image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
      },
      {
        name: "Chronicles of seb",
        price: 28.73,
        image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
      },
      {
        name: "test",
        price: 28.73,
        image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
      },
    ],
  });

  expect(getSnapshot(list)).toMatchSnapshot();

  expect(states).toMatchSnapshot();
});

it("can add new items - 2", () => {
  const list = WishList.create();
  const patches = [];
  onPatch(list, (patch) => {
    patches.push(patch);
  });

  list.add({
    name: "Chronicles of seb",
    price: 28.73,
    image: " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
  });

  list.items[0].changeName("test");

  expect(patches).toMatchSnapshot();
});
