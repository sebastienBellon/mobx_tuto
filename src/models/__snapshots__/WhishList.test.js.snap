// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items - 2 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
      "name": "Chronicles of seb",
      "price": 28.73,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "test",
  },
]
`;

exports[`can add new items 1`] = `
Object {
  "items": Array [
    Object {
      "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
      "name": "Chronicles of seb",
      "price": 28.73,
    },
    Object {
      "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
      "name": "Chronicles of seb",
      "price": 28.73,
    },
    Object {
      "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
      "name": "test",
      "price": 28.73,
    },
  ],
}
`;

exports[`can add new items 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
        "name": "Chronicles of seb",
        "price": 28.73,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
        "name": "Chronicles of seb",
        "price": 28.73,
      },
      Object {
        "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
        "name": "Chronicles of seb",
        "price": 28.73,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
        "name": "Chronicles of seb",
        "price": 28.73,
      },
      Object {
        "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
        "name": "Chronicles of seb",
        "price": 28.73,
      },
      Object {
        "image": " https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL",
        "name": "test",
        "price": 28.73,
      },
    ],
  },
]
`;
